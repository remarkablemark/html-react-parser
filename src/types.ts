import type { DomHandlerOptions } from 'domhandler';
import type { DOMNode } from 'html-dom-parser';
import type { ParserOptions } from 'htmlparser2';
import type { JSX, ReactNode } from 'react';

export interface HTMLReactParserOptions {
  htmlparser2?: ParserOptions & DomHandlerOptions;

  library?: {
    /* eslint-disable @typescript-eslint/no-explicit-any */
    cloneElement: (
      element: JSX.Element,
      props?: object,
      ...children: any[]
    ) => JSX.Element;

    createElement: (
      type: any,
      props?: object,
      ...children: any[]
    ) => JSX.Element;

    isValidElement: (element: any) => boolean;

    [key: string]: any;
    /* eslint-enable @typescript-eslint/no-explicit-any */
  };

  replace?: (
    domNode: DOMNode,
    index: number,
  ) => JSX.Element | string | null | boolean | object | void;

  transform?: (
    reactNode: ReactNode,
    domNode: DOMNode,
    index: number,
  ) => JSX.Element | string | null | void;

  trim?: boolean;
}
