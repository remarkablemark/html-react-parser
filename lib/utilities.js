var styleToJS = require('style-to-js').default;

/**
 * Swap key with value in an object.
 *
 * @param {object} obj - The object.
 * @param {Function} [override] - The override method.
 * @returns - The inverted object.
 */
function invertObject(obj, override) {
  if (!obj || typeof obj !== 'object') {
    throw new TypeError('First argument must be an object');
  }

  var isOverridePresent = typeof override === 'function';
  var overrides = {};
  var result = {};

  for (var key in obj) {
    var value = obj[key];

    if (isOverridePresent) {
      overrides = override(key, value);
      if (overrides && overrides.length === 2) {
        result[overrides[0]] = overrides[1];
        continue;
      }
    }

    if (typeof value === 'string') {
      result[value] = key;
    }
  }

  return result;
}

var RESERVED_SVG_MATHML_ELEMENTS = new Set([
  'annotation-xml',
  'color-profile',
  'font-face',
  'font-face-src',
  'font-face-uri',
  'font-face-format',
  'font-face-name',
  'missing-glyph'
]);

/**
 * Check if a given tag is a custom component.
 *
 * @see {@link https://github.com/facebook/react/blob/v16.6.3/packages/react-dom/src/shared/isCustomComponent.js}
 *
 * @param {string} tagName - The name of the html tag.
 * @param {object} props - The props being passed to the element.
 * @returns {boolean} - Whether tag is custom component.
 */
function isCustomComponent(tagName, props) {
  if (tagName.indexOf('-') === -1) {
    return props && typeof props.is === 'string';
  }
  // These are reserved SVG and MathML elements.
  // We don't mind this whitelist too much because we expect it to never grow.
  // The alternative is to track the namespace in a few places which is convoluted.
  // https://w3c.github.io/webcomponents/spec/custom/#custom-elements-core-concepts
  if (RESERVED_SVG_MATHML_ELEMENTS.has(tagName)) {
    return false;
  }
  return true;
}

// styleToJSOptions
var STYLE_TO_JS_OPTIONS = { reactCompat: true };

/**
 * Sets style prop.
 *
 * @param {null|undefined|string} style
 * @param {object} props
 */
function setStyleProp(style, props) {
  if (style === null || style === undefined) {
    return;
  }
  try {
    props.style = styleToJS(style, STYLE_TO_JS_OPTIONS);
  } catch (err) {
    props.style = {};
  }
}

// Taken from
// https://github.com/facebook/react/blob/cae635054e17a6f107a39d328649137b83f25972/packages/react-dom/src/client/validateDOMNesting.js#L213
var ELEMENTS_WITH_NO_TEXT_CHILDREN = new Set([
  'tr',
  'tbody',
  'thead',
  'tfoot',
  'colgroup',
  'table',
  'head',
  'html',
  'frameset'
]);

/**
 * Checks if the given node can contain text nodes
 *
 * @param {DomElement} node - Node.
 * @returns - Whether node can contain text nodes.
 */
function canTextBeChildOfNode(node) {
  return !ELEMENTS_WITH_NO_TEXT_CHILDREN.has(node.name);
}

/**
 * Returns the first argument as is.
 *
 * @param {any} arg - The argument to be returned.
 * @returns {any} The input argument `arg`.
 */
function returnFirstArg(arg) {
  return arg;
}

module.exports = {
  ELEMENTS_WITH_NO_TEXT_CHILDREN: ELEMENTS_WITH_NO_TEXT_CHILDREN,
  invertObject: invertObject,
  isCustomComponent: isCustomComponent,
  setStyleProp: setStyleProp,
  canTextBeChildOfNode: canTextBeChildOfNode,
  returnFirstArg: returnFirstArg
};
