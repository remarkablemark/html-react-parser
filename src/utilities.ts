import { version } from 'react';
import StyleToJS from 'style-to-js';
import type { Element } from 'html-dom-parser';

import type { Props } from './attributes-to-props';

const RESERVED_SVG_MATHML_ELEMENTS = new Set([
  'annotation-xml',
  'color-profile',
  'font-face',
  'font-face-src',
  'font-face-uri',
  'font-face-format',
  'font-face-name',
  'missing-glyph',
] as const);

type ReservedSvgMathmlElements =
  typeof RESERVED_SVG_MATHML_ELEMENTS extends Set<infer T> ? T : never;

/**
 * Check if a tag is a custom component.
 *
 * @see {@link https://github.com/facebook/react/blob/v16.6.3/packages/react-dom/src/shared/isCustomComponent.js}
 *
 * @param tagName - Tag name.
 * @param props - Props passed to the element.
 * @returns - Whether the tag is custom component.
 */
export function isCustomComponent(
  tagName: string,
  props?: Record<PropertyKey, string>,
): boolean {
  if (!tagName.includes('-')) {
    return Boolean(props && typeof props.is === 'string');
  }

  // These are reserved SVG and MathML elements.
  // We don't mind this whitelist too much because we expect it to never grow.
  // The alternative is to track the namespace in a few places which is convoluted.
  // https://w3c.github.io/webcomponents/spec/custom/#custom-elements-core-concepts
  if (RESERVED_SVG_MATHML_ELEMENTS.has(tagName as ReservedSvgMathmlElements)) {
    return false;
  }

  return true;
}

const styleOptions = {
  reactCompat: true,
} as const;

/**
 * Sets style prop.
 *
 * @param style - Inline style.
 * @param props - Props object.
 */
export function setStyleProp(style: string, props: Props): void {
  if (typeof style !== 'string') {
    return;
  }

  if (!style.trim()) {
    props.style = {};
    return;
  }

  try {
    props.style = StyleToJS(style, styleOptions);
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
  } catch (error) {
    props.style = {};
  }
}

/**
 * @see https://reactjs.org/blog/2017/09/08/dom-attributes-in-react-16.html
 */
export const PRESERVE_CUSTOM_ATTRIBUTES = Number(version.split('.')[0]) >= 16;

/**
 * @see https://github.com/facebook/react/blob/cae635054e17a6f107a39d328649137b83f25972/packages/react-dom/src/client/validateDOMNesting.js#L213
 */
export const ELEMENTS_WITH_NO_TEXT_CHILDREN = new Set([
  'tr',
  'tbody',
  'thead',
  'tfoot',
  'colgroup',
  'table',
  'head',
  'html',
  'frameset',
] as const);

type ElementsWithNoTextChildren =
  typeof ELEMENTS_WITH_NO_TEXT_CHILDREN extends Set<infer T> ? T : never;

/**
 * Checks if the given node can contain text nodes
 *
 * @param node - Element node.
 * @returns - Whether the node can contain text nodes.
 */
export const canTextBeChildOfNode = (node: Element) =>
  !ELEMENTS_WITH_NO_TEXT_CHILDREN.has(node.name as ElementsWithNoTextChildren);

/**
 * Returns the first argument as is.
 *
 * @param arg - The argument to be returned.
 * @returns - The input argument `arg`.
 */
export const returnFirstArg = (arg: any) => arg;
