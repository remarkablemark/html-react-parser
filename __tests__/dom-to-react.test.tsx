import htmlToDOM from 'html-dom-parser';

import { type DOMNode, Element, type HTMLReactParserOptions } from '../src';
import domToReact from '../src/dom-to-react';
import * as utilities from '../src/utilities';
import { html, svg } from './data';
import { render } from './helpers';

describe('domToReact', () => {
  it.each([
    ['comment', html.comment],
    ['doctype', html.doctype],
  ])('skips %s', (type, value) => {
    expect(domToReact(htmlToDOM(value))).toEqual([]);
  });

  it('converts "text" to "text"', () => {
    const text = 'text';
    expect(domToReact(htmlToDOM(text))).toBe(text);
  });

  it('converts single DOM node to React', () => {
    const reactElement = domToReact(htmlToDOM(html.single));
    expect(reactElement).toMatchSnapshot();
  });

  it('converts multiple DOM nodes to React', () => {
    const reactElements = domToReact(
      htmlToDOM(html.multiple),
    ) as React.JSX.Element[];
    reactElements.forEach((reactElement, index) => {
      expect(reactElement.key).toBe(String(index));
    });
    expect(reactElements).toMatchSnapshot();
  });

  it('converts <textarea> correctly', () => {
    // https://reactjs.org/docs/forms.html#the-textarea-tag
    const reactElement = domToReact(htmlToDOM(html.textarea));
    expect(reactElement).toMatchSnapshot();
  });

  it('does not escape <script> content', () => {
    const reactElement = domToReact(htmlToDOM(html.script));
    expect(reactElement).toMatchSnapshot();
  });

  it('does not escape <style> content', () => {
    const reactElement = domToReact(htmlToDOM(html.style));
    expect(reactElement).toMatchSnapshot();
  });

  it('does not have `children` for void elements', () => {
    const reactElement = domToReact(htmlToDOM(html.img)) as React.JSX.Element;
    expect(reactElement.props.children).toBe(undefined);
  });

  it('does not throw an error for void elements', () => {
    const reactElements = domToReact(htmlToDOM(html.void));
    expect(() => {
      render(<div>{reactElements}</div>);
    }).not.toThrow();
  });

  it('skips doctype and comments', () => {
    const reactElements = domToReact(
      htmlToDOM(html.doctype + html.single + html.comment + html.single),
    ) as React.JSX.Element[];
    expect(reactElements).toHaveLength(2);
    expect(reactElements[0].key).toBe('1');
    expect(reactElements[1].key).toBe('3');
    expect(reactElements).toMatchSnapshot();
  });

  it('converts SVG element with viewBox attribute', () => {
    const reactElement = domToReact(
      htmlToDOM(svg.simple, { lowerCaseAttributeNames: false }),
    );
    expect(reactElement).toMatchSnapshot();
  });

  it('converts custom element with attributes', () => {
    const reactElement = domToReact(htmlToDOM(html.customElement));
    expect(reactElement).toMatchSnapshot();
  });
});

describe('library option', () => {
  const React = require('react');
  const Preact = require('preact');

  it('converts with React by default', () => {
    const reactElement = domToReact(htmlToDOM(html.single));
    expect(React.isValidElement(reactElement)).toBe(true);
    expect(Preact.isValidElement(reactElement)).toBe(false);
    expect(reactElement).toEqual(React.createElement('p', {}, 'foo'));
  });

  it('converts with Preact', () => {
    const parsedElement = domToReact(htmlToDOM(html.single), {
      library: Preact,
    });
    const preactElement = Preact.createElement('p', {}, 'foo');
    expect(React.isValidElement(parsedElement)).toBe(false);
    expect(Preact.isValidElement(parsedElement)).toBe(true);
    // remove `__v` key since it's causing test equality to fail
    // @ts-expect-error Property '__v' does not exist on type '...roperty '__v' does not exist on type 'string'.
    delete parsedElement.__v;
    delete preactElement.__v;
    expect(parsedElement).toEqual(preactElement);
  });
});

describe('replace option', () => {
  it.each([undefined, null, 0, 1, true, false, {}])(
    'does not replace for invalid return value %p',
    (value) => {
      const reactElement = domToReact(htmlToDOM('<br>'), {
        replace: () => value,
      }) as React.JSX.Element;
      expect(reactElement).toEqual(<br />);
    },
  );

  it('does not set key for a single node', () => {
    const reactElement = domToReact(htmlToDOM(html.single), {
      replace: () => <div />,
    }) as React.JSX.Element;
    expect(reactElement.key).toBe(null);
  });

  it('does not modify keys if they are already set', () => {
    const reactElements = domToReact(
      htmlToDOM(html.single + html.customElement),
      {
        replace(domNode) {
          const element = domNode as Element;

          if (element.name === 'p') {
            return <p>replaced foo</p>;
          }

          if (element.name === 'custom-element') {
            return (
              // @ts-expect-error Property 'custom-button' does not exist on type 'JSX.IntrinsicElements'.
              <custom-button
                key="myKey"
                class="myClass"
                custom-attribute="replaced value"
              />
            );
          }
        },
      },
    ) as React.JSX.Element[];

    expect(reactElements[0].key).toBe('0');
    expect(reactElements[1].key).toBe('myKey');
  });

  it('replaces with children', () => {
    const options: HTMLReactParserOptions = {
      replace(domNode) {
        if (domNode instanceof Element) {
          return domToReact(domNode.children as DOMNode[], options);
        }
      },
    };
    const reactElement = domToReact(htmlToDOM('<div>test</div>'), options);
    expect(reactElement).toEqual(<div>test</div>);
  });

  it('passes index as the 2nd argument', () => {
    const reactElement = domToReact(htmlToDOM('<li>one</li><li>two</li>'), {
      replace(domNode, index) {
        expect(typeof index).toBe('number');
      },
    });
    expect(reactElement).toHaveLength(2);
  });
});

describe('transform option', () => {
  it('can wrap all elements', () => {
    const reactElement = domToReact(htmlToDOM(html.list), {
      transform: (reactNode, domNode, index) => {
        return <div key={index}>{reactNode}</div>;
      },
    }) as React.JSX.Element;

    expect(reactElement.key).toBe('0');
    expect(reactElement.props.children.props.children[0].key).toBe('0');
    expect(reactElement.props.children.props.children[1].key).toBe('1');
    expect(reactElement).toMatchSnapshot();
  });
});

describe('domToReact', () => {
  describe('when React >=16', () => {
    it('preserves unknown attributes', () => {
      const reactElement = domToReact(htmlToDOM(html.customElement));
      expect(reactElement).toMatchSnapshot();
    });
  });

  describe('when React <16', () => {
    const { PRESERVE_CUSTOM_ATTRIBUTES } = utilities;

    beforeAll(() => {
      // @ts-expect-error Cannot assign to 'PRESERVE_CUSTOM_ATTRIBUTES' because it is a read-only property.
      utilities.PRESERVE_CUSTOM_ATTRIBUTES = false;
    });

    afterAll(() => {
      // @ts-expect-error Cannot assign to 'PRESERVE_CUSTOM_ATTRIBUTES' because it is a read-only property.
      utilities.PRESERVE_CUSTOM_ATTRIBUTES = PRESERVE_CUSTOM_ATTRIBUTES;
    });

    it('removes unknown attributes', () => {
      const reactElement = domToReact(htmlToDOM(html.customElement));
      expect(reactElement).toMatchSnapshot();
    });
  });
});
