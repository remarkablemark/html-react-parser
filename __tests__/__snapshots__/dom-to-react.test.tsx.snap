// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`domToReact converts <textarea> correctly 1`] = `
<textarea
  defaultValue="foo"
/>
`;

exports[`domToReact converts LaTeX 1`] = `
[
  <span
    className="math"
  >
    \\left(\\right)\\rD\\rightarrow\\reals\\ni
  </span>,
  "
",
]
`;

exports[`domToReact converts SVG element with viewBox attribute 1`] = `
<svg
  id="foo"
  viewBox="0 0 512 512"
>
  Inner
</svg>
`;

exports[`domToReact converts custom element with attributes 1`] = `
<custom-element
  class="myClass"
  custom-attribute="value"
  style={
    {
      "OTransition": "all .5s",
      "lineHeight": "1",
    }
  }
/>
`;

exports[`domToReact converts multiple DOM nodes to React 1`] = `
[
  <p>
    foo
  </p>,
  <p>
    bar
  </p>,
]
`;

exports[`domToReact converts single DOM node to React 1`] = `
<p>
  foo
</p>
`;

exports[`domToReact does not escape <script> content 1`] = `
<script
  dangerouslySetInnerHTML={
    {
      "__html": "alert(1 < 2);",
    }
  }
/>
`;

exports[`domToReact does not escape <style> content 1`] = `
<style
  dangerouslySetInnerHTML={
    {
      "__html": "body > .foo { color: #f00; }",
    }
  }
/>
`;

exports[`domToReact skips doctype and comments 1`] = `
[
  <p>
    foo
  </p>,
  <p>
    foo
  </p>,
]
`;

exports[`domToReact when React <16 removes unknown attributes 1`] = `
<custom-element
  className="myClass"
  style={
    {
      "OTransition": "all .5s",
      "lineHeight": "1",
    }
  }
/>
`;

exports[`domToReact when React >=16 preserves unknown attributes 1`] = `
<custom-element
  class="myClass"
  custom-attribute="value"
  style={
    {
      "OTransition": "all .5s",
      "lineHeight": "1",
    }
  }
/>
`;

exports[`transform option can wrap all elements 1`] = `
<div>
  <ol>
    <div>
      <li>
        <div>
          One
        </div>
      </li>
    </div>
    <div>
      <li
        value="2"
      >
        <div>
          Two
        </div>
      </li>
    </div>
  </ol>
</div>
`;
